### PR TITLE
8334670: SSLSocketOutputRecord buffer miscalculation

### PR DESCRIPTION
**Pull request by ascarpino**
Hi,

I need a review to change the a fragment buffer size miscalculation error.   This appears when there are large handshake messages and hasn't been observed during application data.  This was found during testing of the NewSessionTicket change in [JDK-8328608](https://bugs.openjdk.org/browse/JDK-8328608).  There is no regression test as the failure hasn't shown to fail every time. 

thanks

Tony

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8334670](https://bugs.openjdk.org/browse/JDK-8334670): SSLSocketOutputRecord buffer miscalculation (**Bug** - P3)

### Reviewers
 * [Daniel Jeliński](https://openjdk.org/census#djelinski) (@djelinski - **Reviewer**)
 * [Sibabrata Sahoo](https://openjdk.org/census#ssahoo) (@sisahoo - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19862/head:pull/19862` \
`$ git checkout pull/19862`

Update a local copy of the PR: \
`$ git checkout pull/19862` \
`$ git pull https://git.openjdk.org/jdk.git pull/19862/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19862`

View PR using the GUI difftool: \
`$ git pr show -t 19862`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19862.diff">https://git.openjdk.org/jdk/pull/19862.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19862#issuecomment-2186914298)

### CODE REVIEW

**Review by djelinski — APPROVED** [review]
LGTM. Thanks for fixing this!

In order to test this fix, we would need to trigger a situation where `count != position`, and it seems to be impossible, because we flush after every single message. We don't need to flush that often; I'll log a ticket for this.

File: src/java.base/share/classes/sun/security/ssl/SSLSocketOutputRecord.java [hunk 1]
```diff
@@ -169,7 +169,8 @@ void encodeHandshake(byte[] source,
             for (int limit = (offset + length); offset < limit;) {
 
                 int remains = (limit - offset) + (count - position);
```
Comment (djelinski) [inline src/java.base/share/classes/sun/security/ssl/SSLSocketOutputRecord.java:171]:
```suggestion
                int remains = (limit - offset);
```

**Review by djelinski — CHANGES_REQUESTED** [review]
Actually, I think this line needs to be changed too.

File: src/java.base/share/classes/sun/security/ssl/SSLSocketOutputRecord.java (hunk 1 above)
Comment (ascarpino) [reply src/java.base/share/classes/sun/security/ssl/SSLSocketOutputRecord.java:171]: I tried this and it caused a lockup in one of the tests.  I see why your think this is the right change, but it isn't proving out in the testing

Comment (djelinski) [reply src/java.base/share/classes/sun/security/ssl/SSLSocketOutputRecord.java:171]: That's very interesting! Which test was it? Was it with or without #19465?

Comment (ascarpino) [reply src/java.base/share/classes/sun/security/ssl/SSLSocketOutputRecord.java:171]: I do run the tests layered on the current #19465, even though this PR is separate.

**Review by sisahoo — COMMENTED** [review]
Is it possible to add a Test?

**Review by djelinski — APPROVED** [review]
Thanks for making the change. LGTM.
> Is it possible to add a Test?

I don't think it is possible; right now we flush after every handshake message. As a result, `count == position` every time we enter this loop, so the change doesn't really change anything. The problem only surfaced in #19465, and only when multiple NewSessionTicket messages exceeding 16KB were sent without flushing.

**Review by sisahoo — APPROVED** [review]

**Pull request closed without merging**